ouroboros-consensus/src/Ouroboros/Consensus/NodeKernel.hs

```diff
@@ -238,178 +238,226 @@ forkBlockForging
-          case eBlkCtx of
-            Right blkCtx -> return blkCtx
-            Left failure -> do
-              trace failure
-              exitEarly
-
-        trace $ TraceBlockContext currentSlot bcBlockNo bcPrevPoint
-
-        -- Get ledger state corresponding to bcPrevPoint
-        --
-        -- This might fail if, in between choosing 'bcPrevPoint' and this call to
-        -- 'getPastLedger', we switched to a fork where 'bcPrevPoint' is no longer
-        -- on our chain. When that happens, we simply give up on the chance to
-        -- produce a block.
-        --
-        -- Also get a snapshot of the mempool that is consistent with that ledger
-        --
-        -- NOTE: It is possible that due to adoption of new blocks the
-        -- /current/ ledger will have changed. This doesn't matter: we will
-        -- produce a block that fits onto the ledger we got above; if the
-        -- ledger in the meantime changes, the block we produce here may or
-        -- may not be adopted, but it won't be invalid.
-        (unticked, tickedLedgerState, mempoolSnapshot) <- do
-          mb <- lift $ getLedgerAndSnapshotFor mempool bcPrevPoint currentSlot
-          case mb of
-            Just x  -> return x
-            Nothing -> do
-              trace $ TraceNoLedgerState currentSlot bcPrevPoint
-              exitEarly
-
-        trace $ TraceLedgerState currentSlot bcPrevPoint
-
-        -- We require the ticked ledger view in order to construct the ticked
-        -- 'ChainDepState'.
-        ledgerView <-
-          case runExcept $ forecastFor
-                           (ledgerViewForecastAt
-                              (configLedger cfg)
-                              (ledgerState unticked))
-                           currentSlot of
-            Left err -> do
-              -- There are so many empty slots between the tip of our chain and
-              -- the current slot that we cannot get an ledger view anymore
-              -- In principle, this is no problem; we can still produce a block
-              -- (we use the ticked ledger state). However, we probably don't
-              -- /want/ to produce a block in this case; we are most likely
-              -- missing a blocks on our chain.
-              trace $ TraceNoLedgerView currentSlot err
-              exitEarly
-            Right lv ->
-              return lv
-
-        trace $ TraceLedgerView currentSlot
-
-        -- Tick the 'ChainDepState' for the 'SlotNo' we're producing a block
-        -- for. We only need the ticked 'ChainDepState' to check the whether
-        -- we're a leader. This is much cheaper than ticking the entire
-        -- 'ExtLedgerState'.
-        let tickedChainDepState :: Ticked (ChainDepState (BlockProtocol blk))
-            tickedChainDepState =
-                tickChainDepState
-                  (configConsensus cfg)
-                  ledgerView
+        res <- ChainDB.withLgrReadLock chainDB $ withEarlyExit $ do
```

Comment:
We should explain why we hold the ledger read lock here.

Reply:
Hm, let me try to add a comment about that.

Reply:
Done